docs/upload_specifications.ADR.md

```diff
@@ -0,0 +1,170 @@
+   * The `README.md` should include a link to rich upload specification documentation on the KBase
+     website once it is developed.
+   * Note: cover booleans which are not intuitive.
+
+These files will reside in the StS repo. As part of the front end effort, some means of
+delivering the templates and instructions to users will be developed.
+
+Currently, for all in scope apps, the inputs are all strings, numbers, or booleans. There are
+no unusual inputs such as grouped parameters or dynamic lookups. Including future upload apps
+with these features in CSV-based upload may require additional engineering.
+
+## User operations
+
+* The user uploads the upload specification files to the staging area along with all the files
+  inluded in the specification. 
+* The user selects the `Upload Specification` type for the specification files.
+  * The user may also select other files in the staging area to include in the upload along
+    with the files listed in the specification.
+    * The user *does not* have to select any files included in the specification.
+* The user clicks `Import Selected`.
+
+As such, a new type must be added to the StS: `upload_specification` with the title
+`Upload Specification`. *Nota bene*: It may be preferable to have the Narrative specify the
+titles to display in the staging area dropdown rather than the StS.
+
+## Narrative Bulk Upload cell operations
+
+When the bulk upload cell sees a `upload_specification` data type, it calls out to the StS
+endpoint detailed below to get the parsed specifications, and then autofills itself with
+those specifications.
+
+## Staging service upload specification endpoint
+
+The StS endpoint responds to an HTTP GET with a file list in a `files` URL parameter. It is
+extremely unlikely that there will be a large enough set of data types that URL length limits
+will be reached so this seems adequate. The StS will respond with the contents of the files
+parsed into a structure that is similar to that of the input spec, mapped by data type:
+
+```
+Headers:
+Authorization: <KBase token>
+
+GET /bulk_specification/?files=<file1.csv[,file2.csv,...]>
+
+Response:
+{"types": {
+    "gff_metagenome": [
+        {"fasta_file": "mygenome.fa",
+         "gff_file": "mygenome.gff3",
+         "genome_name": "mygenomeobject",
+         "source": null,
+         "release": null,
+         "genetic_code: 11,
+         "generate_missing_genes: 0
+         },
+        {"fasta_file": "mygenome2.fa",
+         "gff_file": "mygenome2.gff3",
+         "genome_name": "mygenomeobject2",
+         "source": "yermumspoo",
+         "release": 30456,
+         "genetic_code: 11,
+         "generate_missing_genes: 1
+         },
+         ...
+    ],
+    "sra_reads": [
+        ...
+    ]
+    ...
+}
+```
+
+The order of the input structures MUST be the same as the order in the input files.
+
+Notably, the service will provide the contents of the files as is and will not perform most error
```

Comment:
Do we need to do any normalization of  strings, booleans, numbers?
Like if the user doesn't provide a value, like species, do they get back
```
{'name': 'mouse'}
```
vs
```
{ 'name' : 'mouse',
'species' : None }
```
Can users mix True/False/0/1, "True", "False"
What about numbers? "0" vs 0
What about "quotation marks"?

Reply:
For the first issue I don't have a strong opinion, but in the example I return null entries.

For the second issue, the service will return whatever's in the file.

Reply:
Hmm, actually maybe we do need to at least convert strings to numbers.

Reply:
Let's see how Pandas loads xSV files. It may just take care of it.

Reply:
Although - presumably the input from HTML text boxes are also strings, so maybe just leaving things as strings is ok. @briehl ?

Reply:
After @briehl 's approval I double checked with him out of band that strings for numbers and booleans are ok